service/src/main/java/bio/terra/workspace/app/controller/ControlledGcpResourceApiController.java

```diff
@@ -537,51 +551,111 @@ private ApiCloneControlledGcpBigQueryDatasetResult fetchCloneBigQueryDatasetResu
+    ApiPrivateResourceUser inputUser = commonFields.getPrivateResourceUser();
+
+    // Shared access has no private user role
+    if (accessScope == AccessScopeType.ACCESS_SCOPE_SHARED) {
+      validateNoInputUser(inputUser);
+      return new PrivateUserRole(false, null, null);
+    }
+
+    // Private access scope
+    switch (managedBy) {
+      case MANAGED_BY_APPLICATION:
+        {
+          // Supplying a user is optional for applications
+          if (inputUser == null) {
+            return new PrivateUserRole(false, null, null);
+          }
+
+          // We have a private user, so make sure the email is present and valid
+          String userEmail = commonFields.getPrivateResourceUser().getUserName();
+          ControllerValidationUtils.validateEmail(userEmail);
+
+          /*
+           TODO: the APPLICATION role does not have read_policies action, so cannot perform
+            this check. We need to give it read_policies.
+          // Validate that the assigned user is a member of the workspace. It must have at least READ action.
+           SamRethrow.onInterrupted(
+               () -> samService.userIsAuthorized(
+                   SamConstants.SAM_WORKSPACE_RESOURCE,
+                   workspaceId.toString(),
+                   SamConstants.SAM_WORKSPACE_READ_ACTION,
+                   userEmail,
+                   userRequest), "validate private user is workspace member");
+           */
+          // Translate the incoming role list into our internal model form
+          // This also validates that the incoming API model values are correct.
+          List<ControlledResourceIamRole> roles =
+              commonFields.getPrivateResourceUser().getPrivateResourceIamRoles().stream()
+                  .map(ControlledResourceIamRole::fromApiModel)
+                  .collect(Collectors.toList());
+          if (roles.isEmpty()) {
+            throw new ValidationException(
+                "You must specify at least one role when you specify PrivateResourceIamRoles");
+          }
+
+          // The legal options for the assigned user of an application is READER
+          // or WRITER. EDITOR is not allowed. We take the "max" of READER and WRITER.
+          var maxRole = ControlledResourceIamRole.READER;
+          for (ControlledResourceIamRole role : roles) {
+            if (role == ControlledResourceIamRole.WRITER) {
+              if (maxRole == ControlledResourceIamRole.READER) {
+                maxRole = role;
+              }
+            } else if (role != ControlledResourceIamRole.READER) {
+              throw new ValidationException(
+                  "For application private controlled resources, only READER and WRITER roles are allowed. Found "
+                      + role.toApiModel());
+            }
+          }
+
+          return new PrivateUserRole(true, userEmail, maxRole);
+        }
+
+      case MANAGED_BY_USER:
```

Comment:
I could move it the function to the service, but I doubt is is interesting to reuse since it operates with API classes. The fundamental choice we made that has it here is constructing the resource objects in the controller. We could move that function to the service and then we would be messing with API classes in the service. I don't see either way as being tidier than the other.

Reply:
I wasn't really thinking of knowledge in terms of the API types, but more the knowledge of the rules for validation. If those could be in one authoritative place, it might make things more robust.